daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala

```diff
@@ -343,9 +345,15 @@ object Node {
     }(recorded, isReplayedBy)
 
   /** Useful in various circumstances -- basically this is what a ledger implementation must use as
-    * a key.
+    * a key. The 'hash' should be used when storing the key as value serialization is not guaranteed
+    * to be stable over time (e.g. a new encoding might be introduced).
     */
-  case class GlobalKey(templateId: Identifier, key: VersionedValue[Nothing])
+  case class GlobalKey private (templateId: Identifier, key: Value[Nothing], hash: Hash)
```

Comment:
Not necessary, but a different approach I could suggest is using a `sealed abstract case class`: https://gist.github.com/tpolecat/a5cb0dc9adeacc93f846835ed21c92d2

Reply:
this language never ceases to amaze me with the amount of footguns it has to offer.